src/realm/db.cpp

```diff
@@ -78,304 +78,192 @@ namespace {
+            data[i].count_live = 0;
+            data[i].count_frozen = 0;
             data[i].current_top = 0;
             data[i].filesize = 0;
-            data[i].next = i + 1;
+            data[i].older = i + 1;
+            data[i].newer = nil;
         }
-        data[new_entries - 1].next = old_pos;
-        data[put_pos.load(std::memory_order_relaxed)].next = entries;
+        data[new_entries - 1].older = freelist;
+        freelist = entries;
         entries = uint32_t(new_entries);
-        // dump();
+    }
+
+    VersionList() noexcept
+    {
+        entries = 0;
+        oldest = newest = nil; // empty
+        freelist = nil;        // empty
+        expand_to(init_readers_size);
     }
 
     static size_t compute_required_space(uint_fast32_t num_entries) noexcept
     {
         // get space required for given number of entries beyond the initial count.
-        // NB: this not the size of the ringbuffer, it is the size minus whatever was
+        // NB: this not the size of the VersionList, it is the size minus whatever was
         // the initial size.
         return sizeof(ReadCount) * (num_entries - init_readers_size);
     }
 
-    uint_fast32_t get_num_entries() const noexcept
+    unsigned int get_num_entries() const noexcept
     {
         return entries;
     }
 
-    uint_fast32_t last() const noexcept
-    {
-        return put_pos.load(std::memory_order_acquire);
-    }
-
-    const ReadCount& get(uint_fast32_t idx) const noexcept
+    ReadCount& get(uint_fast32_t idx) noexcept
     {
         return data[idx];
     }
 
-    const ReadCount& get_last() const noexcept
+    ReadCount& get_oldest() noexcept
     {
-        return get(last());
+        return get(oldest);
     }
 
-    // This method re-initialises the last used ringbuffer entry to hold a new entry.
-    // Precondition: This should *only* be done if the caller has established that she
-    // is the only thread/process that has access to the ringbuffer. It is currently
-    // called from init_versioning(), which is called by DB::open() under the
-    // condition that it is the session initiator and under guard by the control mutex,
-    // thus ensuring the precondition.
-    // It is most likely not suited for any other use.
-    ReadCount& reinit_last() noexcept
+    ReadCount& get_newest() noexcept
     {
-        ReadCount& r = data[last()];
-        // r.count is an atomic<> due to other usage constraints. Right here, we're
-        // operating under mutex protection, so the use of an atomic store is immaterial
-        // and just forced on us by the type of r.count.
-        // You'll find the full discussion of how r.count is operated and why it must be
-        // an atomic earlier in this file.
-        r.count.store(0, std::memory_order_relaxed);
-        return r;
+        return get(newest);
     }
-
-    const ReadCount& get_oldest() const noexcept
+    ReadCount& allocate_entry()
```

Comment:
I think all functions in this class except for `purge` will never throw. It may be worth slapping the `noexcept` on there so that it is clear that we don't need to worry about exception safety in them.